### DISCUSSION TITLE
Exec format error

### DISCUSSION THREAD

**Discussion by christoph-wijns** [ipxe/ipxe#564] [Q&A] [answered]
Guys,

You've already helped me out with iseq problem - I don't want to be lazy but that's the next 'problem'

I'm trying to load Blancco ISO

![2022-01-17_13h10_56](https://user-images.githubusercontent.com/75322374/149767418-74bc507a-65fc-4e12-a2ba-74cf395ddc80.png)

Trying on VMWare - not tested on physical machine

Thanks in advanced

**Comment by NiKiZe** [comment]
Are you perhaps in efi mode and trying to use memdisk which is a pcbios binary?
"Could not boot: Exec format error (http://ipxe.org/2e008081)"
If you go thru the linked paged properly, I'm quite confident you will find the answer.

By the way, don't expect that you will be able to boot isos this way.

**Reply by christoph-wijns** [reply]
Very right - legacy BIOS works flawlessly

Stupid me ...

Can you please guide me trough the process to boot ISO in EFI ?

I´m not asking to do the job for me but a blog/URL would be nice

By the way iPXE loves speed :)

Many thanks

**Reply by NiKiZe** [reply]
It usually don't work. Not even memdisk in pcbios. But you can try to just use sanboot. Will update with link to other posts if I get the time and can find it.

**Reply by christoph-wijns** [reply]
I've extracted blancco iso and I found GRUB_EFI.CFG
Content
  linuxefi /arch/boot/x86_64/vmlinuz archisobasedir=arch archisolabel=BLANCCO copytoram cow_spacesize=2G memtest=00 vmalloc=400M loglevel=0 flr libata.allow_tpm=1
  initrdefi /arch/boot/intel_ucode.img /arch/boot/config.img /arch/boot/x86_64/archiso.img

I'm trying this one in menu
kernel http://My_iPXE_Server/ipxe/Boot/iso/vmlinuz
initrd http://My_iPXE_Server/ipxe/Boot/iso/ARCHISO.IMG

Error
![blancco_1](https://user-images.githubusercontent.com/75322374/149896204-7cc4cabe-51d6-41ff-9f41-e5d7ae431ea7.png)

Am I on right track ?

**Reply by NiKiZe** [reply]
The main thing you are missing is initrd=archiso.img so the kernel knows which initrd to use.
(Searching for "ipxe efi not syncing: VFS: Unable to mount root" you can for example find https://forum.ipxe.org/showthread.php?tid=10770)

Based on
```
linuxefi /arch/boot/x86_64/vmlinuz archisobasedir=arch archisolabel=BLANCCO copytoram cow_spacesize=2G memtest=00 vmalloc=400M loglevel=0 flr libata.allow_tpm=1
initrdefi /arch/boot/intel_ucode.img /arch/boot/config.img /arch/boot/x86_64/archiso.img
```

And that you have http://My_iPXE_Server/ipxe/Boot/iso/
lets create http://My_iPXE_Server/ipxe/Boot/iso/boot.ipxe to make things a bit simpler (and use relative paths)
So boot.ipxe would be:
```bash
#!ipxe
kernel vmlinuz initrd=intel_ucode.img initrd=config.img initrd=archiso.img archisobasedir=arch archisolabel=BLANCCO copytoram cow_spacesize=2G memtest=00 vmalloc=400M loglevel=0 flr libata.allow_tpm=1
initrd intel_ucode.img
initrd config.img
initrd archiso.img
boot
```

`kernel vmlinuz` will download `vmlinuz` from the same location as `boot.ipxe`.
And the same goes for the initrds
The other `initrd=` on the vmlinuz line is so that the kernel knows which files to load. this is required in efi mode but not in pcbios.

Now there is `archisobasedir=arch archisolabel=BLANCCO` which is not clear what it does. but I assume that you have some type of squashfs file on the disk. the system needs to know how to use this somehow and that is different between different distributions so I will not try to cover that. But I will point you to https://gentoo.ipxe.se which has some samples and explanations on the topic, I think that https://gentoo.ipxe.se/altcombine.ipxe would cover most of the explanation.

**Reply by mcb30** [reply]
Using `initrd=initrd.magic` should also work (with no need to then include multiple `initrd=...` arguments), as described in https://ipxe.org/cmd/initrd

**Comment by christoph-wijns** [comment]
NiKiZe & mcb30 you guys are rock solid !!

initrd=initrd.magic did thé trick

I'm getting somewhere with 

#!ipxe
kernel http://ipxe_server/ipxe/Boot/blancco/ARCH/BOOT/X86_64/VMLINUZ initrd=initrd.magic copytoram cow_spacesize=2G memtest=00 vmalloc=400M loglevel=0 flr libata.allow_tpm=1 archisobasedir=arch archiso_http_srv=http://ipxe_server/ipxe/Boot/blancco/ ip=:::::eth0:dhcp initrd=ARCHISO.IMG
initrd http://ipxe_server/ipxe/Boot/blancco/ARCH/BOOT/X86_64/ARCHISO.IMG
boot

but it loops ...

https://user-images.githubusercontent.com/75322374/149945347-e217dba4-a6ce-4414-91ef-307a500572a9.mp4

**Comment by christoph-wijns** [comment, marked as answer]
And **thank** you iPXE - mcb30 - NiKiZe

#!ipxe
kernel http://ipxe-server/ipxe/Boot/blancco/ARCH/BOOT/X86_64/VMLINUZ initrd=initrd.magic copytoram cow_spacesize=2G memtest=00 vmalloc=400M loglevel=0 flr libata.allow_tpm=1 archisobasedir=arch archiso_http_srv=http://ipxe-server/ipxe/Boot/blancco/ ip=dhcp BOOTIF=${netX/mac} initrd=ARCHISO.IMG
initrd http://ipxe-server/ipxe/Boot/blancco/ARCH/BOOT/intel_ucode.img 
initrd http://ipxe-server/ipxe/Boot/blancco/ARCH/BOOT/config.img
initrd http://ipxe-server/ipxe/Boot/blancco/ARCH/BOOT/X86_64/ARCHISO.IMG
boot

![blancco_3](https://user-images.githubusercontent.com/75322374/149947564-ddf2f98f-5a6f-4acd-b5af-215d0f60509e.png)
